IOPath interface and IOPathSpec implementation

cmnbroad opened the pull request:
An interface and extensible concrete implementation for paths specifiers backed by URIs. The hard part is the names.

yfarjoun commented:
@cmnbroad could you elaborate a little on what problem this solves? what will be easier to do using this interface?

cmnbroad commented:
Yeah, of course. The idea is to have a single interface and concrete implementation for use in existing and new htsjdk interfaces, for specifying input or output resources that are on a local file system; a remote file system such as GCS, or some other resource; such as an htsget or refget uri/specifier.

The constructor accepts a string that can be either an absolute or relative local file name, or a URI, including scheme and query parameters. The (IOPaths) are always backed by a URI with a protocol scheme (`file://` if nothing else) that can be queried to determine how to handle the resource, even if one wasn't provided to the constructor. The idea is to provide a single currency (for new code at least), and help eliminate the proliferation of `File`/`Path`/`String` method overloads and interconversions, and eliminate ambiguities around handling path names that contain embedded characters such as space or '#' (like [this one](https://github.com/samtools/htsjdk/pull/1487)), since these are always backed by a normalized, encoded URI.

Callers can provide subclassed implementations for resolving paths that require special handling (GATK for example uses this to provide to a `Path` for both gs:// and hdfs:// access).

This will be the first step toward support for versioned, side-by-side file formats and https://github.com/samtools/htsjdk/issues/1482, though I think it will be useful even with the existing classes.

cmnbroad commented on src/main/java/htsjdk/io/IOPath.java:
```diff
@@ -0,0 +1,192 @@
+package htsjdk.io;
```
Yeah, the first commit has the GATK versions. The only changes were the package and class names, exceptions, and a handful of test result changes to reflect the different file systems available in htjsjdk.

cmnbroad commented on src/main/java/htsjdk/io/IOPathSpec.java:
```diff
@@ -0,0 +1,263 @@
+ *         opaque_part   = uric_no_slash *uric
+ *         uric_no_slash = unreserved | escaped | ";" | "?" | ":" | "@" | "&" | "=" | "+" | "$" | ","
+ */
+public class IOPathSpec implements IOPath, Serializable {
```
HtsjdkPath is fine with me.

cmnbroad commented:
I updated the class name to `HtsjdkPath`, leaving the interface name as IOPath, and changed some test variable/method/provider names to reflect that change.

cmnbroad commented:
Any other community comments/input before we move forward with this ?

lbergelson commented:
No one is chiming in so lets merge this.
